### PR TITLE
update(test-case): add a test case to directly connect local-emulator.

### PR DESCRIPTION
**Pull request by xmano**
Add testing library infrastructure #3247 

Currently I am trying to configure the tests to directly exercise the local emulator.

Testing with just single spec functions.specs.ts

Connection is made.
Verifying with other test case to write and read a document.

Debugging the failure :

`
    > 25 |     await expect(db.doc('testCollection/testDoc').get())                                                          
         |           ^                                                                                                       
      26 |           .resolves.toBe({testStr});                                                                              
      27 |   });                                                                                                             
      28 |                                                                                                                   
                                                                                                                             
      at expect (node_modules/expect/build/index.js:138:15)                                                                  
      at libs/testing/src/lib/firebase/functions.spec.ts:25:11                                                               
      at step (libs/testing/src/lib/firebase/functions.spec.ts:33:23)                                                        
      at Object.next (libs/testing/src/lib/firebase/functions.spec.ts:14:53)                                                 
      at libs/testing/src/lib/firebase/functions.spec.ts:8:71                                                                
      at new ZoneAwarePromise (node_modules/zone.js/dist/zone.js:913:33)                                                     
      at Object.<anonymous>.__awaiter (libs/testing/src/lib/firebase/functions.spec.ts:4:12)                                 
      at libs/testing/src/lib/firebase/functions.spec.ts:18:40                                                               
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:386:30)                      
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/dist/proxy.js:117:43)                 
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone.js:385:36)                      
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone.js:143:47)                                         
                                                                                                                             
  console.error node_modules/zone.js/dist/zone.js:690                                                                        
    Unhandled Promise rejection:                                                                                             
    Null value error. for 'get' @ L21 ; Zone: <root> ; Task: setTimeout ; Value: Error [FirebaseError]:                      
    Null value error. for 'get' @ L21                                                                                        
`

### CODE REVIEW

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 1]
```diff
@@ -1,8 +1,31 @@
 import { initFunctionsTestMock } from './functions';
+import * as firebaseTesting from '@firebase/testing'
 import { firestore, initializeApp } from 'firebase-admin';
 
+const project_ID = "blockframes-mano";
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:5]: You don't want to give your real project ID here. You want a random value for each projectId I think, else I fear you might have concurrency issues between tests that are using the same projectId

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 2]
```diff
@@ -1,8 +1,31 @@
...
+
 describe('firebase testing library', () => {
   test.todo('Hello World!');
+
+  it('should access emulator directly', async () => {
+    const db = firebaseTesting.initializeTestApp({
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:11]: Where are you saying that you're using the emulator for this test ?

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 3]
```diff
@@ -1,8 +1,31 @@
 import { initFunctionsTestMock } from './functions';
+import * as firebaseTesting from '@firebase/testing'
 import { firestore, initializeApp } from 'firebase-admin';
 
+const project_ID = "blockframes-mano";
+
 describe('firebase testing library', () => {
   test.todo('Hello World!');
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:8]: Please remove that.

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 4]
```diff
@@ -1,16 +1,54 @@
...
-    const test = 5;
-    await db
-      .collection('test')
-      .doc('test')
-      .set({ test });
-    await expect(db.doc('test/test').get()).resolves.toBe(5);
+
+  describe('Direct access to emulator', () => {
+    it.skip('should access emulator directly', async () => {
+      //TODO : This is for reference
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:9]: In the project, whenever we add a TODO, if you don't solve it in the same PR, you need to create an issue related to this TODO & add the id of the issue in the comment. For example : 
```typescript
// @todo(#1234) Make it reusable
```

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 5]
```diff
@@ -1,16 +1,54 @@
...
+      //TODO: Debug the test failure.                
+      await expect(db.doc('testCollection/testDoc').get())
+            .resolves.toBe({testStr});
+    });
+  })
+
+  describe('Access emulator through library', () => {
+    let db;
+
+    beforeAll(() => {
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:33]: I think you want that to be `beforeEach` to make sure the DB is clean between each test.
Or you can clean it with `afterEach`. Maybe it's faster 🤔

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 6]
```diff
@@ -1,16 +1,54 @@
...
+    }) 
+
+    it('should be able to access the emulator', async () => {
+      const testDoc = await db.collection("testCollection").doc("testDoc");
+      firebaseTesting.assertSucceeds(testDoc.get());    
+    });
+
+    it('should be able to read the document', async () => {
+      const testStr = "blockframes";
+      const testDoc = db.collection("testCollection")
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:45]: As you're going to write that a lot I would recommand you to name collection "col", document "doc", subcollection "subCol", subdocument : "subdoc": 
```typescript
const doc = db.doc('col/doc').set({ value });
```

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 7]
```diff
@@ -1,16 +1,54 @@
...
+                      .doc("testDoc").set({testStr});
+      //TODO: Debug the test failure.      
+      //Expectation failure : Received promise rejected instead of resolved          
+      await expect(db.doc('testCollection/testDoc').get())
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:49]: I think you should do something like : 
```typescript
const doc = await db.doc('testCollection/testDoc').get();
expect(doc.data()).toEqual({ testStr });
```
You cannot use `toBe` as `toBe` check the memory reference and not the deep equality.

Comment (george43g) [reply libs/testing/src/lib/firebase/functions.spec.ts:49]: `await expect(db.doc.get).resolves....(use deep equality compare with written object)`

File: jest.config.js [hunk 8]
```diff
@@ -15,14 +15,15 @@ module.exports = {
   transform: {
     '^.+\\.(ts|js|html)$': 'ts-jest'
   },
-  testEnvironment: 'jest-environment-jsdom-thirteen',
+  testEnvironment: 'node',
```
Comment (GrandSchtroumpf) [inline jest.config.js:18]: I'm not sure it's a good idea to put it here. What will happen when we run angular unit test ?

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 9]
```diff
@@ -1,16 +1,17 @@
-import { initFunctionsTestMock } from './functions';
+﻿import * as firebaseTesting from '@firebase/testing'
 import { firestore, initializeApp } from 'firebase-admin';
 
 describe('firebase testing library', () => {
-  test.todo('Hello World!');
   it.skip('should be able to access the emulator', async () => {
-    const firebaseTest = initFunctionsTestMock();
+    //Init test DB
+    const projectId = String(Math.random());
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:7]: Why are you not using the code from George ?

File: libs/testing/jest.config.js [hunk 10]
```diff
@@ -1,9 +1,14 @@
...
   name: 'testing',
   preset: '../../jest.config.js',
   transform: {
     '^.+\\.[tj]sx?$': 'ts-jest'
   },
-  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'html'],
+  //Note: testEnvironment may not be appropriate for some app tests
+  testEnvironment: 'node',
+  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'html', 'json'],
+  testPathIgnorePatterns: [
```
Comment (GrandSchtroumpf) [inline libs/testing/jest.config.js:10]: Remove if not needed.

File: jest.config.js [hunk 11]
```diff
@@ -23,6 +23,7 @@ module.exports = {
     'jest-preset-angular/build/HTMLCommentSerializer.js'
   ],
   testPathIgnorePatterns: [
+    'node_modules/',
```
Comment (GrandSchtroumpf) [inline jest.config.js:26]: remove if not needed.

File: libs/testing/src/lib/firebase/functions.spec.ts [hunk 12]
```diff
@@ -1,16 +1,16 @@
...
-    const test = 5;
-    await db
-      .collection('test')
-      .doc('test')
-      .set({ test });
-    await expect(db.doc('test/test').get()).resolves.toBe(5);
+    const docRef = db.collection('testCollection').doc('testDoc');
+    const testData = { name: 'blockframes' };
+    await docRef.set(testData);
+    const doc = await db.doc('testCollection/testDoc').get();
```
Comment (GrandSchtroumpf) [inline libs/testing/src/lib/firebase/functions.spec.ts:13]: You can do `const doc = await docRef.get()` to not repeat yourself.

File: jest.config.js (hunk 8 above)
Comment (xmano) [reply jest.config.js]: removed and moved to lib config

**Comment by xmano** [comment]
Closes #3247

**Pull request merged**